### ISSUE TITLE
Responder should check `END_SESSION` request attributes

### ISSUE THREAD

**Issue by steven-bellock** [DMTF/libspdm#1953]
The Responder does not validate the Requester's `END_SESSION` request attributes. In particular if the Requester shall not set `SPDM_END_SESSION_REQUEST_ATTRIBUTES_PRESERVE_NEGOTIATED_STATE_CLEAR` if the Responder's `CACHE_CAP=0`.

https://github.com/DMTF/libspdm/blob/8ada4786f07447fbf5f28c21f7a3114c2917fa1f/library/spdm_responder_lib/libspdm_rsp_end_session.c#L70

**Comment by PavelPu123** [comment]
Hi Steven
Regarding "end_session_attributes" - I don't see it being used by the responder. As I understand it should be used to reset the VCA state. Am I missing something?
Thanks.

**Comment by steven-bellock** [comment]
Integrator would need to check that through the `libspdm_session_state_callback_func` right before the session ends. For `Negotiated State Preservation Indicator`, right now it's up to them to manage the cached `VCA`, and settings derived from it. We may come up with a better `CACHE_CAP` interface / flow in the future if folks actually start using it.

As an aside I don't know why `Negotiated State Preservation Indicator` even exists. If the Responder supports `CACHE_CAP`, and the Requester wants to re-negotiate `VCA`, it just needs to send `GET_VERSION`.

**Comment by steven-bellock** [comment]
> Am I right?

No, the current connection's `VCA` is still valid after `Negotiated State Preservation Indicator` has been set. For example, if a Requester sends `END_SESSION` with `Negotiated State Preservation Indicator` set, it can subsequently issue a `CHALLENGE` request, and it would need the `VCA` messages to calculate the `M1/M2` transcript.

**Comment by PavelPu123** [comment]
From Table 77 (Page 146) in DSP0274_1.2.1:

"If the Responder supports Negotiated State caching ( CACHE_CAP=1 ), **the Responder shall also clear the
Negotiated State as part of session termination.** If there is no Negotiated State to be cleared due to a
previous END_SESSION request message with this field set to 1, this field shall be ignored. If the Responder
does not support Negotiated State caching ( CACHE_CAP=0 ), this field shall be ignored."

From DSP2058_1.2.0 (page 57):

351 12.2 Secured Message error handling
352 If an error occurs during the Session Handshake phase or if an error happens during a secure session, the Negotiated State is preserved. **The Negotiated State is preserved through errors unless the Requester sends an END_SESSION request with Negotiated State Preservation Indicator=1 to terminate the session or sends a GET_VERSION request to reset the session.**

It seems that in case of "Negotiated State Preservation Indicator=1" the state should be cleared. Isn't it?

**Comment by steven-bellock** [comment]
That section needs `cached` in front of `Negotiated State`.
>If the Responder supports Negotiated State caching ( CACHE_CAP=1 ), the Responder shall preserve the **cached** Negotiated State.

>If the Responder supports Negotiated State caching ( CACHE_CAP=1 ), the Responder shall also clear the **cached** Negotiated State as part of session termination. If there is no **cached** Negotiated State to be cleared due to a previous END_SESSION request message with this field set to 1, this field shall be ignored. If the Responder does not support Negotiated State caching ( CACHE_CAP=0 ), this field shall be ignored.

I'll file an issue against the specification.

Also, after reading that again, it looks like this issue can be modified since
>If the Responder does not support Negotiated State caching ( CACHE_CAP=0 ), this field shall be ignored.

Ie, if the Responder doesn't support caching then it shouldn't be checking this field. I'll add a comment to the code.

**Comment by PavelPu123** [comment]
"I'll file an issue against the specification." - Great. Thanks.

But how about my above question? ("It seems that in case of "Negotiated State Preservation Indicator=1" the state should be cleared. Isn't it?")

And how can I quote some previous reply in my reply?

**Comment by steven-bellock** [comment]
>But how about my above question? ("It seems that in case of "Negotiated State Preservation Indicator=1" the state should be cleared. Isn't it?") 

The cached negotiated state should be cleared, but libspdm does not maintain the cached negotiated state. That's (currently) up to the Integrator. `Negotiated State Preservation Indicator` should probably also be directly in the `spdm_context` since it affects the connection and not just a single session. I can file a separate issue against libspdm for that.

>And how can I quote some previous reply in my reply?

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

**Comment by PavelPu123** [comment]
>but libspdm does not maintain the cached negotiated state.

But it does clear the cache state when GET_VERSION is received here:
https://github.com/DMTF/libspdm/blob/6684f43117784b12e25e665df65d1d122540f11f/library/spdm_responder_lib/libspdm_rsp_version.c#L60
Which is according to the above spec mentions, so why shouldn't is handle the other part of this:
>The Negotiated State is preserved through errors unless the Requester sends an END_SESSION request with Negotiated State Preservation Indicator=1 to terminate the session or sends a GET_VERSION request to reset the session.

>should probably also be directly in the spdm_context
It seems so.

**Comment by steven-bellock** [comment]
>The Negotiated State is preserved through errors **unless the Requester sends an END_SESSION request with Negotiated State Preservation Indicator=1 to terminate the session or sends a GET_VERSION request to reset the session.**

@bhenning10 The highlighted portion of the sentence can be removed from the whitepaper, no? `Negotiated State Preservation Indicator` does not affect the currently Negotiated State, only the cached Negotiated State if the Responder's `CACHE_CAP=1`.

**Comment by bhenning10** [comment]
I see what you're saying. errors aren't the same thing as END_SESSION, so the first and second parts of the sentence aren't talking about the same thing.
What about a wording that starts the sentence with
The `Negotiated State` is preserved through `END_SESSION` requests, unless...

**Comment by steven-bellock** [comment]
> I'm guessing the word "cached" should also be added in this sentence before "Negotiated State" (in the same manner you proposed above):

The only things that affect the `cached Negotiated State` is `GET_VERSION` and `Negotiated State Preservation Indicator` so yes, it can be re-worded to reflect that.

>like it does in case of GET_VERSION (which I mentioned above). It seems like a small fix and it makes sense.

Explain how the Requester and Responder could perform a successful `CHALLENGE / CHALLENGE_AUTH` immediately after the session has ended, with `Negotiated State Preservation Indicator=1`, when the Responder has cleared / reset its message A (`VCA`) transcript.

**Comment by PavelPu123** [comment]
>Explain how the Requester and Responder could perform a successful CHALLENGE / CHALLENGE_AUTH immediately after the session has ended, with Negotiated State Preservation Indicator=1, when the Responder has cleared / reset its message A (VCA) transcript.

As I understand from this sentence:
>The Negotiated State is preserved through errors unless the Requester sends an END_SESSION request with Negotiated State Preservation Indicator=1 to terminate the session or sends a GET_VERSION request to reset the session.
in case this flag is set, this is similar to sending GET_VERSION, which clears the cached negotiated state, so in both cases you wouldn't be able to perform CHALLENGE / CHALLENGE_AUTH until a new state is negotiated. Maybe I missed something in the spec about this? 
In any case it seems that the logic of this flag isn't defined well in specifications.

**Comment by steven-bellock** [comment]
See https://github.com/DMTF/libspdm/issues/1953#issuecomment-1520394702 where `cached` will be put in front in front of `Negotiated State`.
